File: generated-code/rust/codecraft/FileReadWrite/src/model/auto_attack.rs
use super::*;

/// Auto attack options
#[derive(Clone, Debug)]
pub struct AutoAttack {
    /// Maximum distance to pathfind
    pub pathfind_range: i32,
    /// List of target entity types to try to attack. If empty, all types but resource are considered
    pub valid_targets: Vec<model::EntityType>,
}

impl trans::Trans for AutoAttack {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.pathfind_range.write_to(writer)?;
        self.valid_targets.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let pathfind_range: i32 = trans::Trans::read_from(reader)?;
        let valid_targets: Vec<model::EntityType> = trans::Trans::read_from(reader)?;
        Ok(Self {
            pathfind_range,
            valid_targets,
        })
    }
}

File: generated-code/php/codecraft/FileReadWrite/Model/AutoAttack.php
<?php

namespace Model {
    require_once 'Model/EntityType.php';
    require_once 'Stream.php';

    /**
     * Auto attack options
     */
    class AutoAttack
    {
        /**
         * Maximum distance to pathfind
         */
        public int $pathfindRange;
        /**
         * List of target entity types to try to attack. If empty, all types but resource are considered
         */
        public array $validTargets;
    
        function __construct(int $pathfindRange, array $validTargets)
        {
            $this->pathfindRange = $pathfindRange;
            $this->validTargets = $validTargets;
        }
    
        /**
         * Read AutoAttack from input stream
         */
        public static function readFrom(\InputStream $stream): AutoAttack
        {
            $pathfindRange = $stream->readInt32();
            $validTargets = [];
            $validTargetsSize = $stream->readInt32();
            for ($validTargetsIndex = 0; $validTargetsIndex < $validTargetsSize; $validTargetsIndex++) {
                $validTargetsElement = \Model\EntityType::readFrom($stream);
                $validTargets[] = $validTargetsElement;
            }
            return new AutoAttack($pathfindRange, $validTargets);
        }
        
        /**
         * Write AutoAttack to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32($this->pathfindRange);
            $stream->writeInt32(count($this->validTargets));
            foreach ($this->validTargets as $element) {
                $stream->writeInt32($element);
            }
        }
    }
}